File: Data Structures & Algorithms/clone-graph/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;
        unordered_map<Node*,Node*> tracking;
        return dfs(node,tracking);
        
    }

    Node* dfs(Node * vert,unordered_map<Node*,Node*> & tracking){
        if (!vert) return nullptr;

        if ( tracking.count(vert)) return tracking[vert];
            Node * newvert= new Node(vert->val);
            tracking[vert]= newvert;
            for (Node * neighbor : vert->neighbors){
                dfs(neighbor, tracking);
                tracking[vert]->neighbors.push_back(tracking[neighbor]);
            }
        return tracking[vert];
    }
};

File: Data Structures & Algorithms/count-connected-components/submission-0.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        if not edges:
            return 0
        graph = defaultdict(list)
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)
        # {0: [1], 
        # 1: [0, 2], 
        # 2: [1], 
        # 3: [4], 
        # 4: [3]
        visited=set()
        count=0 
        def dfs(curr):
            if curr not in visited:
                visited.add(curr)
            for neighbor in graph[curr]:
                if neighbor in visited: 
                    continue
                else:
                    dfs(neighbor)
        for node in range(n):
            if node not in visited:
                count+=1
                dfs(node)
        return count